src/station_cmd.cpp

```diff
@@ -2220,34 +2241,87 @@ static CommandCost RemoveRoadStop(TileIndex tile, DoCommandFlag flags, int repla
 }
 
 /**
- * Remove bus or truck stops.
- * @param flags Operation to perform.
- * @param tile Northernmost tile of the removal area.
- * @param width Width of the removal area.
- * @param height Height of the removal area.
- * @param stop_type Type of stop (bus/truck).
- * @param remove_road Remove roads of drive-through stops?
- * @return The cost of this operation or an error.
+ * Remove a road waypoint
+ * @param tile TileIndex been queried
+ * @param flags operation to perform
+ * @param replacement_spec_index replacement spec index to avoid deallocating, if < 0, tile is not being replaced
+ * @return cost or failure of operation
  */
-CommandCost CmdRemoveRoadStop(DoCommandFlag flags, TileIndex tile, uint8_t width, uint8_t height, RoadStopType stop_type, bool remove_road)
+CommandCost RemoveRoadWaypointStop(TileIndex tile, DoCommandFlag flags, int replacement_spec_index)
 {
-	if (stop_type >= ROADSTOP_END) return CMD_ERROR;
-	/* Check for incorrect width / height. */
-	if (width == 0 || height == 0) return CMD_ERROR;
-	/* Check if the first tile and the last tile are valid */
-	if (!IsValidTile(tile) || TileAddWrap(tile, width - 1, height - 1) == INVALID_TILE) return CMD_ERROR;
-	/* Bankrupting company is not supposed to remove roads, there may be road vehicles. */
-	if (remove_road && (flags & DC_BANKRUPT)) return CMD_ERROR;
+	Waypoint *wp = Waypoint::GetByTile(tile);
 
-	TileArea roadstop_area(tile, width, height);
+	if (_current_company != OWNER_WATER) {
+		CommandCost ret = CheckOwnership(wp->owner);
+		if (ret.Failed()) return ret;
+	}
+
+	/* don't do the check for drive-through road stops when company bankrupts */
+	if (!(flags & DC_BANKRUPT)) {
+		CommandCost ret = EnsureNoVehicleOnGround(tile);
+		if (ret.Failed()) return ret;
+	}
+
+	const RoadStopSpec *spec = GetRoadStopSpec(tile);
+
+	if (flags & DC_EXEC) {
+		/* Update company infrastructure counts. */
+		for (RoadTramType rtt : _roadtramtypes) {
+			RoadType rt = GetRoadType(tile, rtt);
+			UpdateCompanyRoadInfrastructure(rt, GetRoadOwner(tile, rtt), -static_cast<int>(ROAD_STOP_TRACKBIT_FACTOR));
+		}
+
+		Company::Get(wp->owner)->infrastructure.station--;
+		DirtyCompanyInfrastructureWindows(wp->owner);
+
+		DeleteAnimatedTile(tile);
+
+		uint specindex = GetCustomRoadStopSpecIndex(tile);
+
+		DeleteNewGRFInspectWindow(GSF_ROADSTOPS, tile.base());
+
+		DoClearSquare(tile);
+
+		wp->rect.AfterRemoveTile(wp, tile);
+
+		wp->RemoveRoadStopTileData(tile);
+		if ((int)specindex != replacement_spec_index) DeallocateSpecFromRoadStop(wp, specindex);
```

Comment:
Why can't `replacement_spec_index` be `uint`?

Reply:
uint cannot represent negative values. See the function header and other uses of replacement_spec_index.